First look: Overture Schema Pydantic port

sethfitz opened the pull request:
It's complete, but it's super rough. Please review for broad smells that we can address generally. This PR is only for follow-up on the threads below at this point. Please review the stack of PRs instead.

sethfitz committed:
Import examples and counterexamples from schema

This includes some fixes made to their validity.

sethfitz committed:
Add complete Pydantic Overture Schema implementation

- Core base classes with extension support and theme-type validation
- Full schema coverage: addresses, base, buildings, divisions, places, transportation
- Multi-package structure with theme-based namespaces, type-specific
  packages, and common code
- Validation using examples and counterexamples

sethfitz committed:
Improve README

sethfitz committed:
Use constraint-based validation

...based on Vic's prototype

sethfitz committed:
Start curating generated JSON Schema

sethfitz committed:
Document model registration

sethfitz committed:
Fully move to constraint-based validation

sethfitz committed:
Fix ruff linter settings deprecation

sethfitz committed:
Create a feedback loop for improving JSON Schema output

sethfitz committed:
Define a feedback loop for JSON Schema generation

sethfitz committed:
Additional constraint validation work and discoveries

sethfitz committed:
Migrate @model_validators to mixin-based constraints

sethfitz committed:
Ignore line length problems

sethfitz committed:
Fix/improve testing infrastructure

sethfitz committed:
Workspace improvements

- add a Makefile-based test runner
- improve workspace/project structure (including better support for
  overlapping namespaces)
- move division-specific constraints out of core validation package

sethfitz committed:
Add rail segment counterexamples to improve coverage

sethfitz committed:
Add support for running test coverage

sethfitz committed:
Remove unused CountryCode models

sethfitz committed:
Clean up address validation

sethfitz committed:
Remove redundant legacy segment validation

vcschapp commented:
First pass cursory comments.

# Big Picture 🌍

1. Good progress, an impressive amount of code done!
2. I'm overwhelmed, on various axes:
    - The diff is too big to comment on, bogs multiple browsers down to a crawl, makes trying to review, comment, *etc.* nigh-impossible/very frustrating.
    - I'm composing this comment in an outside editor just to get away from the GH UI.
    - The sheer surface area of change also I think makes trying to grok it all pretty tough.
    - I think we should try for a "vertical slice iteration" (I hate to use the term, but ...) that presents a diff for a smaller scope first so we can get more aligned on the key conceptual parts before expanding to breadth. I would advocate for divisions + transportation.
3. I really want to use [the approach I sketched](https://github.com/vcschapp/overture-schema-pydantic/) for the base feature and geometry, especially with regard to isolating the GeoJSON to the JSON Schema returned by `BaseModel.model_json_schema()` and not allowing it to creep into the model anywhere else.
4. I really want to see the scoping concept, which I worked very hard on building up and unifying, carefully reflected in the model.
    - Either an annotation-driven approach or an inheritance-driven approach would be nice here, with my preference being annotation.
    - Some comments sprinkled throughout the diff.
5. The examples/counterexamples should validate the migration, but IMO once we hit the validated baseline, we only use them to regression test the JSON Schema output.
6. We should discuss the purpose and pros/cons of the type/theme registry. I'm worried it will give us an upkeep headache and an illusion of control that doesn't do a lot of good in practice.

# Packaging 📦

## Top-Level Packages

tl;dr, I think I royally goofed [in my sketch suggestion to split the packaging all the way down to the feature type level](https://github.com/OvertureMaps/rfd/discussions/22).

It seemed smart when you only look at two features but with the whole schema in front of us, the number of packages is overwhelming also. It's just too granular, there's too many of 'em captain!

As well, the example of the two theme-level packages `overture-schema-buildings-common` and `overture-schema-theme-buildings`  highlights that we have two opposing purposes for a theme-level package:

1. Aggregate all the types for the theme.
2. Provide shared common code that's unique to the theme but shared among theme features.

Obviously if all the packages are split down to the feature type level, this becomes a major challenge.

On reflection, and changing my previous opinion from the RFD sketch, I think we should consolidate at the theme level and just run with a package dependency tree like:

```
overture-schema-core
├── overture-schema-theme-addresses
├── overture-schema-theme-base
├── overture-schema-theme-buildings
├── overture-schema-theme-divisions
├── overture-schema-theme-places
├── overture-schema-theme-transportation
...
└── overture-schema-codegen
    └── overture-conductor-cli

```

This does mean that an isolated breaking change to, say, `building` could have an unwanted versioning effect on `building_part`, but at least it wouldn't affect any other theme like `addresses` or `places`. I think we have to compromise to get a package structure that is grokkable.

## Modules

We have the ability to organize packages into modules so to get the sub-package organization level we desire.

I'd like to rough out the modules we want, especially in the `core` package but also in one or two theme-level packages to see what we arrive at in terms of general principles.


# In the Weeds 🌱

## Terminology Discussion

- We should avoid too many "general terms" *e.g.* base, core, common, *etc.*
   - Especially when mixed together because when you have 3 different generalities and nothing specific it just seems weird.
   - We should stay away from "base" as a general term because it's already an Overture theme.
   - I have a preference for `core` as the main "stuff that's shared" term.
- Personal aesthetic opinion: I prefer  `Feature` and `Division` style better than `OvertureFeature` and `DivisionFeature`.
- We should decide whether we want to consistently apply the suffix `Constraint` to constraint classes.
    - I was inconsistent about it in my first sketch, *e.g.* [`GeometryTypeConstraint` ](https://github.com/vcschapp/overture-schema-pydantic/blob/main/src/overture_schema_pydantic/geometry.py#L27) is a main constraint class with the suffix but [`MinItems`](https://github.com/vcschapp/overture-schema-pydantic/blob/main/src/overture_schema_pydantic/constraint.py#L35) is also a constraint without it.
    - My aesthetic preference FWIW is to drop the suffix assuming we can do so fairly consistently.
- We should try to leave the `Container` suffix (*e.g.* `CartographyContainer`) behind. That terminology was very much tied to the JSON Schema approach.

## Removing Vestiges of the Old Order

1. I'm interested in making `theme` a method of `Feature`, not an actual data-containing field.
2. I believe the `ext_` properties were a noble attempt but it's not clear they work to do anything useful.
   - IMO we should seize the opportunity of the Pydantic port to delete this feature from the schema, as the new-and-better Real Extension Support will blow it away anyway.
   - Pydantic models are permissive about unrecognized fields by default, so I propose we just lean into that default except in the places (like `names.common`) where allowing unrecognized things is conceptually broken.

## Decorators/Mixins

I like the decorators like `@required_if` but have some questions and thoughts:

1. If we use this approach, do they really need to come with a global registry as in `mixin.py` line 14? Can't we just edit the class in-place, for example replacing a member function with a wrapped version that adds the constraint layer?
2. Is this going to be sufficiently discoverable for other non-Pydantic purposes such as our code generation framework?
    - Ideally the code generation framework lives in a different package and can "see" everything it needs to be able to generate the full range of useful validation behavior.
    - Ideally the code generation can traverse the Python class hierarchy and find enough meaningful semantic metadata to generate the full range of needed behavior. This will be difficult if we're just focused on changing Pydantics validation behavior.
    - I think an annotation-based approach might be better (and more consistent), because the annotations are visible via `__class__.__orig_bases__` and `typing.get_origin` and `typoing.get_args`.
    - We could have the decorators add the annotations?

sethfitz committed:
Remove unused (and non-functional) model-level constraints

sethfitz committed:
Fix connector test

sethfitz committed:
Replace validation rules that operated outside Pydantic

sethfitz committed:
Document the scoping system

sethfitz committed:
ScopeContainer → Scope

sethfitz committed:
Remove Feature suffix

sethfitz commented:
> 2. I'm overwhelmed, on various axes:
>    
>    * The diff is too big to comment on, bogs multiple browsers down to a crawl, makes trying to review, comment, _etc._ nigh-impossible/very frustrating.
>    * The sheer surface area of change also I think makes trying to grok it all pretty tough.
>    * I think we should try for a "vertical slice iteration" (I hate to use the term, but ...) that presents a diff for a smaller scope first so we can get more aligned on the key conceptual parts before expanding to breadth. I would advocate for divisions + transportation.

Agreed. Once I get to the point of having something "complete", I want to refactor the git history and produce a stack of PRs that can be reviewed independently (by a broader set of people, since the pain will be more acute there).

> 3. I really want to use [the approach I sketched](https://github.com/vcschapp/overture-schema-pydantic/) for the base feature and geometry, especially with regard to isolating the GeoJSON to the JSON Schema returned by `BaseModel.model_json_schema()` and not allowing it to creep into the model anywhere else.

I'll tackle this soon (I also like the sketch). I wanted to minimize the delta between JSON Schema + Pydantic to start with, but I've got a baseline now.

> 4. I really want to see the scoping concept, which I worked very hard on building up and unifying, carefully reflected in the model.
>    
>    * Either an annotation-driven approach or an inheritance-driven approach would be nice here, with my preference being annotation.

Did you have any initial thoughts on what the annotation-driven approach might look like yet, or is that still rolling around?

>    * Some comments sprinkled throughout the diff.

👍 

> 5. The examples/counterexamples should validate the migration, but IMO once we hit the validated baseline, we only use them to regression test the JSON Schema output.

Agreed.

> 6. We should discuss the purpose and pros/cons of the type/theme registry. I'm worried it will give us an upkeep headache and an illusion of control that doesn't do a lot of good in practice.

I'm not sold on it yet. I noticed that I was hard-coding a bunch of theme and type constants in a way that was going to make it difficult for extensions to reuse some of the validation + JSON Schema generation infrastructure without extra boilerplate (which was getting error-prone in the tests). The boilerplate might not be so bad though.

> # Packaging 📦
> ## Top-Level Packages
> tl;dr, I think I royally goofed [in my sketch suggestion to split the packaging all the way down to the feature type level](https://github.com/OvertureMaps/rfd/discussions/22).
> 
> It seemed smart when you only look at two features but with the whole schema in front of us, the number of packages is overwhelming also. It's just too granular, there's too many of 'em captain!
> 
> As well, the example of the two theme-level packages `overture-schema-buildings-common` and `overture-schema-theme-buildings` highlights that we have two opposing purposes for a theme-level package:
> 
> 1. Aggregate all the types for the theme.
> 2. Provide shared common code that's unique to the theme but shared among theme features.
> 
> Obviously if all the packages are split down to the feature type level, this becomes a major challenge.
> 
> On reflection, and changing my previous opinion from the RFD sketch, I think we should consolidate at the theme level and just run with a package dependency tree like:
> 
> ```
> overture-schema-core
> ├── overture-schema-theme-addresses
> ├── overture-schema-theme-base
> ├── overture-schema-theme-buildings
> ├── overture-schema-theme-divisions
> ├── overture-schema-theme-places
> ├── overture-schema-theme-transportation
> ...
> └── overture-schema-codegen
>     └── overture-conductor-cli
> ```
> 
> This does mean that an isolated breaking change to, say, `building` could have an unwanted versioning effect on `building_part`, but at least it wouldn't affect any other theme like `addresses` or `places`. I think we have to compromise to get a package structure that is grokkable.

I don't have a problem with the proliferation / granularity of the packages (because it allows focused changes at the cost of recognizing when common structures should be shared across a theme), but I'd also be ok with theme-level consolidation (esp. since it solves the opposing purposes idea in a simpler way) while keeping the namespaces/modules laid out in the same way.

> ## Modules
> We have the ability to organize packages into modules so to get the sub-package organization level we desire.
> 
> I'd like to rough out the modules we want, especially in the `core` package but also in one or two theme-level packages to see what we arrive at in terms of general principles.

Sure. I know it's probably hard to see, but do you have specific objections so the current namespace/model layout, or do you see it as a later thing we should consider (if not act upon) before we're "done"?

> # In the Weeds 🌱
> ## Terminology Discussion
> * We should avoid too many "general terms" _e.g._ base, core, common, _etc._
>   
>   * Especially when mixed together because when you have 3 different generalities and nothing specific it just seems weird.
>   * We should stay away from "base" as a general term because it's already an Overture theme.
>   * I have a preference for `core` as the main "stuff that's shared" term.

Agreed. 👍 

> * Personal aesthetic opinion: I prefer  `Feature` and `Division` style better than `OvertureFeature` and `DivisionFeature`.

Agreed.

> * We should decide whether we want to consistently apply the suffix `Constraint` to constraint classes.

🤷 I keep having to steer Claude away from using `@field_validator` and `@model_validator` and generally found that referring to the validation system as "constraint-based validation" helped to disambiguate.

>   
>   * I was inconsistent about it in my first sketch, _e.g._ [`GeometryTypeConstraint` ](https://github.com/vcschapp/overture-schema-pydantic/blob/main/src/overture_schema_pydantic/geometry.py#L27) is a main constraint class with the suffix but [`MinItems`](https://github.com/vcschapp/overture-schema-pydantic/blob/main/src/overture_schema_pydantic/constraint.py#L35) is also a constraint without it.
>   * My aesthetic preference FWIW is to drop the suffix assuming we can do so fairly consistently.

I'll add taking a closer look to my queue.

> * We should try to leave the `Container` suffix (_e.g._ `CartographyContainer`) behind. That terminology was very much tied to the JSON Schema approach.

What was the history of "container"? Reading the JSON Schema, it feels like the terminology was used to create reusable concepts within the constraints of the language.

> ## Removing Vestiges of the Old Order
> 1. I'm interested in making `theme` a method of `Feature`, not an actual data-containing field.

Does this matter? If people were using the Pydantic models directly (how Pydantic is traditionally used), that makes sense to me, but since this is more about modeling structure, that doesn't seem valuable.

> 2. I believe the `ext_` properties were a noble attempt but it's not clear they work to do anything useful.
>    
>    * IMO we should seize the opportunity of the Pydantic port to delete this feature from the schema, as the new-and-better Real Extension Support will blow it away anyway.

Agreed on `ext_` fields, though their presence in the counterexamples is helpful (though their use could probably be replaced with YAML comments at this point). Let's remove it before we're "done".

>    * Pydantic models are permissive about unrecognized fields by default, so I propose we just lean into that default except in the places (like `names.common`) where allowing unrecognized things is conceptually broken.

I disagree. I've found that rejecting unrecognized fields has been valuable, as features may have successfully validated due to a typo or similar. That said, we need to make sure that it's easy for extension authors to add fields.

> 
> ## Decorators/Mixins
> I like the decorators like `@required_if` but have some questions and thoughts:
> 
> 1. If we use this approach, do they really need to come with a global registry as in `mixin.py` line 14? Can't we just edit the class in-place, for example replacing a member function with a wrapped version that adds the constraint layer?

This was necessary for testing purposes, as multiple uses (against classes with the same names) leaked between test cases. I'll take a closer look, but this may also be related to how `Annotated` doesn't work at the class level (including inheriting from `Annotated` classes).

> 2. Is this going to be sufficiently discoverable for other non-Pydantic purposes such as our code generation framework?

I believe so, though that inclination derives from being able to modify the JSON Schema output in useful ways (a Pydantic purpose though).

>    * Ideally the code generation framework lives in a different package and can "see" everything it needs to be able to generate the full range of useful validation behavior.
>    * Ideally the code generation can traverse the Python class hierarchy and find enough meaningful semantic metadata to generate the full range of needed behavior. This will be difficult if we're just focused on changing Pydantics validation behavior.
>    * I think an annotation-based approach might be better (and more consistent), because the annotations are visible via `__class__.__orig_bases__` and `typing.get_origin` and `typing.get_args`.
>    * We could have the decorators add the annotations?

Maybe? My goal is to be able to reflect on classes and see everything from the code generation package. I may need to do something else to attach visible metadata (like annotations), but what I've learned so far is that this needs to be driven by the decorators (since annotating classes prevents validations from being registered with Pydantic).

sethfitz committed:
Standardize on Python >=3.10

sethfitz committed:
Fix/remove transportation rules

sethfitz committed:
Use YAML 1.2 Loader

YAML 1.2 removes yes/no boolean-ish values, which we use in our
examples.

sethfitz committed:
Standardize on [dependency-groups]

sethfitz committed:
Simplify example testing + fix linting

sethfitz committed:
Merge `overture-schema-transportation-rules` into `overture-schema-transportation-common`

sethfitz committed:
Remove disabled example reporting

sethfitz committed:
Adopt the Geometry type

sethfitz committed:
Add typed markers

sethfitz committed:
GeoJSON/Shapely/WKB/WKT support

sethfitz committed:
Flatten the model structure + related fixes

GeoJSON inputs and outputs are still supported, but the model is flat.
This surfaced a few issues in the process which we fixed along the way.

sethfitz committed:
Avoid generating duplicate test targets

sethfitz committed:
Modernize type annotations to match our minimum Python version (3.10)

sethfitz committed:
Extract serde and mode-related tests

sethfitz committed:
Use Python >= 3.10

sethfitz committed:
Allow entire example directories to be disabled

sethfitz committed:
Merge overture-schema-address-type with overture-schema-addresses-theme

sethfitz committed:
Merge bathymetry-type with base-theme

sethfitz committed:
Merge infrastructure-type with base-theme

sethfitz committed:
Merge land-cover-type with base-theme

sethfitz committed:
Merge land-type with base-theme

sethfitz committed:
Sync uv.lock

sethfitz committed:
Merge land-use-type with base-theme

sethfitz committed:
Merge water-type with base-theme

sethfitz committed:
Merge base-common with base-theme

sethfitz committed:
Merge building-part-type with buildings-theme

sethfitz committed:
Merge building-type with buildings-theme

sethfitz committed:
Sync Python requirement between all packages

sethfitz committed:
Merge buildings-common with buildings-theme

sethfitz committed:
Merge division-area-type with divisions-theme

sethfitz committed:
Merge division-boundary-type with divisions-theme

sethfitz committed:
Merge division-type with divisions-theme

sethfitz committed:
Merge division-common with divisions-theme

sethfitz committed:
Merge place-type with places-theme

sethfitz committed:
Merge connector-type with transportation-theme

sethfitz committed:
Merge segment-type with transportation-theme

sethfitz committed:
Merge transportation-common with transportation-theme

sethfitz committed:
Reconcile core's common.py

sethfitz committed:
Extract CartographyContainer

sethfitz committed:
Remove redundant py.typed marker

vcschapp commented:
> >1. I'm interested in making `theme` a method of Feature, not an actual data-containing field.
> 
> Does this matter? If people were using the Pydantic models directly (how Pydantic is traditionally used), that makes sense to me, but since this is more about modeling structure, that doesn't seem valuable.

I think the Pydantic models will be used. It's not a pan-Overture priority, but even within Overture they'd be super useful in the transportation combobulator, which is all written in Python. Outside Overture, given Python's status as a go-to tool for quick prototyping and data analysis (and it's status in the uni. system/with new undergrads!) I do believe the models will be used. So I don't think they're *just* for expressing the schema model.....

Getting rid of `theme` matters - if it's redundant, this is the opportunity to do it. OTOH, if we find some use for it in the extension system then maybe it gets a stay of execution.

vcschapp commented:
> > I really want to see the scoping concept, which I worked very hard on building up and unifying, carefully reflected in the model.
> >
> > Either an annotation-driven approach or an inheritance-driven approach would be nice here, with my preference being annotation.

> Did you have any initial thoughts on what the annotation-driven approach might look like yet, or is that still rolling around?

I remember mocking it up somewhere but now can't recall where. Here are the different ways it *could* be modeled.

### Decorators

```python
@scoped(Scope.GeometricRange, Scope.Temporal, Scope.Heading)
class LaneAccessRestrictions: BaseModel:
    access: AccessMode
```

### Annotations

```python
class LaneAccessRestrictions(Annotated[BaseModel, Scoped(Scope.GeometricRange, Scope.Temporal, Scope.Heading)]):
    access: AccessMode
```

### Inheritance

```python
class LaneAccessRestrictions(BaseModel, scope.GeometricRange, scope.Temporal, scope.Heading):
    access: AccessMode
```

I have a preference for the first two options, as the scoping seems to me to be more naturally modeled as a "mix-in function" rather than a polymorphic behavior.

Of the two options, the first one seems cleaner, and it might be more viable in the sense that Python itself has a driver for running decorators...

vcschapp commented:
> > We have the ability to organize packages into modules so to get the sub-package organization level we desire.
> > 
> > I'd like to rough out the modules we want, especially in the core package but also in one or two theme-level packages to see what we arrive at in terms of general principles.
> 
> Sure. I know it's probably hard to see, but do you have specific objections so the current namespace/model layout, or do you see it as a later thing we should consider (if not act upon) before we're "done"?

The modules I *know* we need:

- `scoping` - Contains all the foundations for scoped properties.

The modules I think we might need:

- `types` - Core types. Everything from `Names` to `Source` and all the primitives. (Maybe this should be split into `scalar` and something like `model` that contains the reusable compound types that are based on `BaseModel`?)
- `constraints` - Maybe, just extra constraints you can annotate stuff with.

The module names I want to avoid:

- `base` - Overlaps with the "base" theme.
- `common` / `shared` / `core` - Too generic and overlaps with the "core" package.

vcschapp commented:
> > - Pydantic models are permissive about unrecognized fields by default, so I propose we just lean into that default except in the places (like `names.common`) where allowing unrecognized things is conceptually broken.
> 
> I disagree. I've found that rejecting unrecognized fields has been valuable, as features may have successfully validated due to a typo or similar. That said, we need to make sure that it's easy for extension authors to add fields.

I can go along with this, as long as we all remember that we're strictly speaking narrowing the schema (by eliminating `ext_` and not replacing it with "permissive" properties such that `ext_` will still be valid.

I think this is OK, personally, as I really think `ext_` was a blind alley - good intentions, but ultimately not useful.

As you say we really just need to make it possible to extend the schema *easily*.
